Deleting spam users who do not complete checkout

dparker1005 opened the pull request:
### All Submissions:

* [x] Have you followed the [Contributing guideline](https://github.com/strangerstudios/paid-memberships-pro/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/strangerstudios/paid-memberships-pro/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
Deleting users who are registered with PMPro but do not complete the checkout process. This is helpful to delete spam accounts that fail the payment step and then bail.

As currently coded, this feature is powered by an hourly cron that deletes spam accounts 24 hours after creation. There is currently a filter to change that window of time and another filter to disable user deletion altogether.

Before releasing, we may want to make UI settings for this feature and we need to figure out how it should be set by default.

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

dparker1005 commented:
There is the `pmpro_spam_pending_delete_user`, but I think that you're right that either a filter to turn off the cron or maybe an advanced setting would be nice. Should consider this before merging. We also need to choose the default time to wait before deleting spam users. Should probably be 2 weeks or so to account for Stripe delayed payment methods

ideadude commented:
Good start. I like the general idea of tagging users at checkout and then removing that tag when checkout is complete. Some notes from our discussion to use when refactoring this:

* Let's use a user taxonomy instead of user meta for performance reasons
* We can maybe use some of the user fields code for the above
* Let's expand on the situations where we will remove the user taxonomy, e.g. creating a post, updating their profile, WC purchase, EDD purchase, others?? so we don't delete users who interacted with non-PMPro stuff after getting logged in by PMPro
* Add an "incomplete membership checkouts" submenu to the users page (that's a mouthful).
* clicking on it adds a param to the URL to filter the WP list table user query by our user taxonomy
* Now folks can manually check users and bulk delete them
* We will develop a paid add on to automate some of this (there are other types of old inactive users we'd want to include in  the users the add on addresses)

dparker1005 commented:
Just finished most of those bullets. The remaining "to do's" are:

- Expanding on situations where we will remove the taxonomy
- Finalizing name for "incomplete membership checkouts" menu item

One other thing on my mind is that when admins filter the users list by these "incomplete" users, there is no visual indication that this is being done (just shows the normal "Users" page UI) and there isn't a way to get a sense of how long ago the user was created. This is problematic because if a user is mid-checkout or if they used a payment method that takes multiple days to complete, the admin does not have a way of knowing this.

ideadude commented:
btw for this I mean a link at the top of the users table... not a link in the sidebar menu.
> Add an "incomplete membership checkouts" submenu to the users page (that's a mouthful).

dparker1005 commented:
Just moved the "Incomplete Membership Checkokuts" option next to the list of roles and added a sortable "User Registered" column when that the "Incomplete Membership Checkouts" option is selected. Should be ready for final review now.

ideadude commented on includes/abandoned-signups.php:
```diff
@@ -0,0 +1,196 @@
+ *
+ * @since TBD
+ */
+function pmpro_abandoned_signups_taxonomy() {
```
I'm thinking it's good to get picky about the naming here. Our first priority is to flag users that never complete checkout so we can delete them after a few days, but in reality this taxonomy can be used later for abandoned cart functionality, which would be really cool.

ideadude replied:
So I'm thinking something more like "pmpro_checkout_user_status" and having terms like "pending", "abandoned", and "completed".

Pending = user created at checkout
Abandoned = maybe updated via cron X days later
Completed = updated when checkout is completed

ideadude replied:
Let's merge pending and abandoned into an "incomplete" status. This avoids the need for a cron, but doesn't have the loaded "abandoned" language that might be problematic if it "gets out" in future updates.

ideadude replied:
Let's merge pending and abandoned into an "incomplete" status. This avoids the need for a cron, but doesn't have the loaded "abandoned" language that might be problematic if it "gets out" in future updates.

ideadude commented on includes/abandoned-signups.php:
```diff
@@ -0,0 +1,196 @@
+	}
+
+	// Add the abandoned signup taxonomy to the user.
+	var_dump( wp_set_object_terms( $user_id, 'abandoned-signup', 'pmpro_abandoned_signup' ) );
```
Note the var_dump here.

ideadude commented on includes/abandoned-signups.php:
```diff
@@ -0,0 +1,196 @@
+	// Remove the abandoned signup taxonomy from the user.
+	wp_remove_object_terms( $user_id, 'abandoned-signup', 'pmpro_abandoned_signup' );
+}
+add_action( 'pmpro_after_checkout', 'pmpro_remove_abandoned_signup_taxonomy' );
```
I think we need to think of some other actions that would complete here. So e.g. something on login (but make sure to skip the login we do at checkout) would be good, because if a user logged in user that user/pass a day or two later that's an indication that they are expecting to continue using that user account.

Similarly, we need to check for posts added with them as the author (could be literal posts or maybe show that they interacted with some plugin via CPT)... or maybe specifically check for WooCommerce checkouts, EDD checkouts, etc. What other actions would indicate "this user expects to keep using this user account" besides completing a PMPro checkout?

ideadude replied:
Let's focus on the original target issue here of spam checkouts and just folks who never come back after failing checkout. What we can do is hook into the "wp" hook and if not pmpro_is_checkout() then consider this a user who was logged in and did *something* other than try to checkout with PMPro and remove the inactive/abandoned/spam checkout taxonomy term (or whatever we end up calling it).

So we'll hook into the pmpro_after_checkout and user_deleted hooks, but add a new callback hooked on wp to check if the user has the term and remove it.

This will then run on wp for every logged in user, but should be a quick enough check that it doesn't slow things down.

ideadude commented on includes/abandoned-signups.php:
```diff
@@ -0,0 +1,196 @@
+ *
+ * @param array $views The views on the Users page.
+ */
+function pmpro_add_users_table_view_abandoned_signups( $views ) {
```
If we have 3 statuses now, we probably want to at least allow filtering by pending as well. I don't think filtering by completed is as useful, but maybe people would expect it.

ideadude replied:
No longer relevant if we go back to 1 or 2 statuses.

ideadude reviewed (commented):
Updated with notes RE changing the taxonomy name to "pmpro_checkout_user_status" and using 3 terms (pending, abandoned, completed).

Also an errant var_dump in there.

dparker1005 commented:
Made all the requested changes. The one thing that we still need to think about before 2.11 releases is the name of the taxonomy in the code to potentially remove the word "abandoned".
